src/java/org/apache/cassandra/index/sai/plan/StorageAttachedIndexSearcher.java

```diff
@@ -142,8 +142,10 @@ public UnfilteredPartitionIterator search(ReadExecutionController executionContr
                          queryContext.shadowedKeysLoopCount(),
                          queryContext.rowsMatched(),
                          currentShadowedKeysCount);
-            // Stop if no new shadowed keys found or if we already got enough rows
-            if (newShadowedKeysCount == 0 || exactLimit <= queryContext.rowsMatched())
+            // Stop if no new shadowed keys found or if we already got enough rows (In this case, we're basically
+            // checking that if all the shadowed keys came from one index, we still got enough non-shadowed rows to
+            // guarantee we have a valid result.)
+            if (newShadowedKeysCount == 0 || exactLimit <= queryContext.softLimit() - currentShadowedKeysCount)
```

Comment:
Before #915, it looked like:

```java
if (newShadowedKeysCount == 0 || exactLimit <= softLimit - newShadowedKeysCount)
```

That was wrong too. We need to use the `currentShadowedKeysCount`, as the test shows.

This change might motivate us to look at resumable searches so that we can query deeper as needed. It also might provide motivation to attribute shadowed keys to a certain source index, since we only really need to search deeper in those graphs.

Reply:
Can you explain why it was wrong? Is `rowsMatched` not tracking the correct number of rows somehow, or is it a problem that we want to terminate sometimes even if we have fewer rows than the exact limit? My line of thinking was that if there are fewer good rows than the exact limit, then eventually we'd also have no more `newShadowedKeys` and this would terminate.

Reply:
Huh, I think this `rowsMatched` logic is flawed. I guess it counts rows even if they were shadowed; so for a deleted row it will count also the original even if it was shadowed by a tombstone. 

So you fixed the termination condition, but the estimation is still using incorrect input data, that's why there are so many rounds.

Reply:
Now when this condition was changed I'm not sure, the `newShadowedKeysCount == 0` condition looks also a bit fishy to me. Can you describe a correct situation where `newShadowedKeysCount == 0` would be true, but the second part after || would be false? 

I believe, we either have enough rows or we don't.

Reply:
> Can you explain why it was wrong?

Shadowed keys can produce incorrect results because they render their source index's top K incomplete, and the `kth + 1` vector on the source index could be greater than other vectors in other indexes. The current algorithm needs to re-query for very basic cases. However, that is more naive than necessary. By keeping the row iterators in score order, we can lazily re-query only the relevant indexes that had too many shadowed rows.